hbt/production/tau.py

```diff
@@ -0,0 +1,317 @@
+# coding: utf-8
+
+"""
+Tau scale factor production.
+"""
+
+import functools
+import itertools
+
+from columnflow.production import Producer, producer
+from columnflow.util import maybe_import
+from columnflow.columnar_util import set_ak_column, flat_np_view, layout_ak_array
+
+ak = maybe_import("awkward")
+np = maybe_import("numpy")
+
+
+# helper
+set_ak_column_f32 = functools.partial(set_ak_column, value_type=np.float32)
+
+
+@producer(
+    uses={
+        "single_triggered", "cross_triggered",
+        "nTau", "Tau.pt", "Tau.eta", "Tau.genPartFlav", "Tau.decayMode",
+    },
+    produces={
+        "tau_weight",
+    } | {
+        f"tau_weight_{unc}_{direction}"
+        for unc, direction in itertools.product(
+            [
+                "jet_dm0", "jet_dm1", "jet_dm10", "e_barrel", "e_endcap",
+                "mu_0p4", "mu_0p4To0p8", "mu_0p8To1p2", "mu_1p2To1p7", "mu_1p7ToInf",
+            ],
+            ["up", "down"],
+        )
+    },
+)
+def tau_weights(self: Producer, events: ak.Array, **kwargs) -> ak.Array:
+    """
+    Producer for tau ID weights. Requires an external file in the config as (e.g.)
+
+    .. code-block:: python
+
+        "tau_sf": ("/afs/cern.ch/user/m/mrieger/public/mirrors/jsonpog-integration-f018adfb/POG/TAU/2017_UL/tau.json.gz", "v1"),  # noqa
+
+    as well as an auxiliary entry in the config to refer to tau tagger name,
+
+    .. code-block:: python
+
+        cfg.x.tau_tagger = "DeepTau2017v2p1"
+
+    that is used to extract correction sets such as "DeepTau2017v2p1VSjet" from the json file.
+
+    Resources:
+    https://twiki.cern.ch/twiki/bin/view/CMS/TauIDRecommendationForRun2?rev=113
+    https://gitlab.cern.ch/cms-nanoAOD/jsonpog-integration/-/blob/849c6a6efef907f4033715d52290d1a661b7e8f9/POG/TAU
+    """
+    # fail when running on data
+    if self.dataset_inst.is_data:
+        raise ValueError("attempt to compute tau weights in data")
+
+    # helper to bring a flat sf array into the shape of taus, and multiply across the tau axis
+    reduce_mul = lambda sf: ak.prod(layout_ak_array(sf, events.Tau.pt), axis=1, mask_identity=False)
+
+    # the correction tool only supports flat arrays, so convert inputs to flat np view first
+    pt = flat_np_view(events.Tau.pt, axis=1)
+    abseta = flat_np_view(abs(events.Tau.eta), axis=1)
+    dm = flat_np_view(events.Tau.decayMode, axis=1)
```

Comment:
Is `Tau.decayMode` automatically `decayModeFindingNewDMs`? (Saw in the TWiki linked above that this is written explicitly, and since I never worked with taus just want to make sure I understand)

Reply:
Yes, the DM in nano always comes from the newest algo. The `decayModeFindingNewDMs` is actually a boolean flag denoting that the DM is actually part of a set that generally counts as "accepted and blessed to work with".